design-proposals/2022-09-02-design-proposal-template.md

```diff
@@ -1,21 +1,155 @@
+
+This is the MapLibre Rendering Modularization Design Plan and the biggest goal is to pave the way for Metal support on iOS. But while we’ve got the hood open, so to speak, we’d also like to fix a few other things. So let’s get to it!
+
+## Goals / Evaluation metrics
+
+This proposal contains details as to the design of a refactored library. If this design is implemented as detailed, we expect the following to be true:
+
+1. Developers should be able to continue using the OpenGL renderer with this refactor.
+
+2. This refactor will allow migration to Metal on IOS with less overhead than a migration without this explicit modularization. In other words, migration will be substantially simplified.
+
+3. The resulting compiled binary from the modularized refactor should not increase more than 5% when compiled.
+
+4. The modularization refactor should be visually identical to the current implementation when using OpenGL. All existing render tests should pass, and developers that continue to use this library with an OpenGL renderer should not experience any visual degradation such as artifacts.
+
+5. After refactor, the startup time for applications build with this library should not increase. This will be measured by the startup time of [this](https://github.com/maplibre/maplibre-gl-native/blob/main/platform/ios/platform/ios/iosapp%20UITests/iosapp_UITests.swift) iOS test app.
+
+6. When implemented with a given style and set of tilesets, the compiled library should not decrease the rendering frame rate.
+
+7. This refactor will not introduce data loading or other bottlenecks that will negatively impact rendering performance.
+
+8. Developers currently using this library will not see any breaking API changes, and will be able to seamlessly upgrade versions.
+
+
+9. Optionally, shader code will no longer be compiled from the [webgl source](https://github.com/maplibre/maplibre-gl-js/tree/main/src/shaders), and should be organized within the `maplibre-gl-native` repo.
+
+10. Shader code will implemented in a modular fashion, such that replacement of any given shader can happen without impacting other shaders
+
+11. Layer rendering logic should be replaceable at the dev level.  All the necessary support should be exposed and a developer should be able to take over representation of any given layer.
+
+12. Developers should be able to associate specific styles with specific layer rendering.  The style sheet should be able to call out the type of representation it would like beyond the default.
+
+13. Developers should be able to add new visual representations fed by existing geometry types.  Data display driven by tile sets, for example.
+
+14. We should be able to drop out unused layer types with a compile flag.
```

Comment:
For the Drawable question, yes, each renderer would have its own subclass of Drawable.  The Metal renderer will draw DrawableMTL and the GL renderer will draw DrawableGL.  The Builders will also have subclasses, so for instance the LineBuilder will have a LineBuilderMTL and a LineBuilderGL.  Some of the logic will be shared in the LineBuilder and some may be shared through helper classes for marshaling buffers.

For the Shaders, that sounds right.  The Layer determines which shader is assigned.  Metal is much more flexible in its shader implementation, so we might not need different shaders for each of those categories. It might work to simply set a Uniform when creating the Drawable.  We'll see how that works out.

For buffers and your point about sharing logic between them, I agree.  But I would put much of that logic in helper classes.  So a helper class for loading points into memory would be useful, for example.  I wouldn't expose the concept of a buffer outside of each SDK's implementation.  They are very similar, but once you get into life cycle issues they can act very differently.

On the subject of specific buffer types, that's more of an issue for OpenGL and I think the GL implementation should keep that logic.  But again I wouldn't enforce that across implementations.  I'll give you an example as to why.  In our own toolkit (WhirlyGlobe-Maply) we've been moving to instancing for a lot of functionality.  Our wide vector implementation now uses this flexible junction instance (a collection of triangles and vertices) which it then instantiates along the centerline of a road.  The shader itself does a lot of the geometry work, much more than we used to in GL and the result is much better.  It uses less memory, is faster to build on the CPU and is much more flexible in the shader.

My point is that we don't want the MapLibre implementation to dictate how the graphics implementation does its work.  With this proposal, the Layer describes what it wants to the LineBuilder, for example.  Then the LineBuilderMTL constructs things as it pleases to implement that request.  The first version will look a lot like the LineBuilderGL, since we'll copy over that work initially.  But the second version may look more like our instancing approach.  Maybe the third version just copies in the centerlines and asks the GPU to do the rest.  That's possible now for the higher class of Metal GPUs and it's clearly where things are going.

These discussions are making it clear that we could explain this better.  The basic idea is: MapLibre describes what it wants and the specific implementation (GL, Metal) implements it.  But we don't dictate how it implements that.

Reply:
Great, no further problem for me.